osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -56,35 +56,75 @@ public async Task<MultiplayerRoom> JoinRoom(long roomId)
+                        }
 
-                    // this is a sanity check to keep *rooms* in a good state.
-                    // in theory the connection clean-up code should handle this correctly.
-                    if (room.Users.Any(u => u.UserID == roomUser.UserID))
-                        throw new InvalidOperationException($"User {roomUser.UserID} attempted to join a room they are already present in.");
+                        // mark the room active - and wait for confirmation of this operation from the database - before adding the user to the room.
+                        await MarkRoomActive(room);
 
-                    // mark the room active - and wait for confirmation of this operation from the database - before adding the user to the room.
-                    await MarkRoomActive(room);
+                        userUsage.Item = new MultiplayerClientState(Context.ConnectionId, CurrentContextUserId, roomId);
+                        room.Users.Add(roomUser);
 
-                    room.Users.Add(roomUser);
+                        await UpdateDatabaseParticipants(room);
 
-                    await UpdateDatabaseParticipants(room);
+                        await Clients.Group(GetGroupId(roomId)).UserJoined(roomUser);
+                        await Groups.AddToGroupAsync(Context.ConnectionId, GetGroupId(roomId));
 
-                    await Clients.Group(GetGroupId(roomId)).UserJoined(roomUser);
-                    await Groups.AddToGroupAsync(Context.ConnectionId, GetGroupId(roomId));
+                        Log($"Joining room {room.RoomID}");
+                    }
+                    catch
+                    {
+                        // if room join failed, we may need to clean up the room usage.
+                        if (room != null)
+                        {
+                            try
+                            {
+                                if (userUsage.Item != null)
+                                {
+                                    // the user was joined to the room, so we can tun the standard leaveRoom method.
+                                    // this will handle closing the room if this was the only user.
+                                    await leaveRoom(userUsage.Item, roomUsage);
+                                }
+                                else
+                                {
+                                    // the room may have been retrieved but failed before the user could join.
+                                    // check whether the room should be closed (may happen if this was the first and only user joining the room)
+                                    if (room.Users.Count == 0)
+                                    {
+                                        await EndDatabaseMatch(room);
+                                        roomUsage.Destroy();
```

Comment:
i'd rather we could put this in the `finally`, but calling `Destroy` twice will result in an exception. Discussed previously silencing this and allow it to noop instead - interested in opinions on this direction still as it may lead to tidier code.